ValveResourceFormat/KeyValues/KVObject.cs

```diff
@@ -0,0 +1,217 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Globalization;
+
+namespace ValveResourceFormat.KeyValues
+{
+    //Different type of value blocks for KeyValues3
+    public enum KVType
+    {
+        OBJECT,
+        ARRAY,
+        BOOLEAN,
+        INTEGER,
+        DOUBLE,
+        STRING,
+        STRING_MULTI,
+        FLAGGED_STRING
+    }
+
+    //Datastructure for a KV Object
+    public class KVObject
+    {
+        public string Key;
+        public Dictionary<String, KVValue> Properties;
+        private bool IsArray;
+        public int Count;
+
+        public KVObject(string name)
+        {
+            Key = name;
+            Properties = new Dictionary<String, KVValue>();
+            IsArray = false;
+            Count = 0;
+        }
+
+        public KVObject(string name, bool isArray)
+            : this(name)
+        {
+            IsArray = isArray;
+        }
+
+        //Add a property to the structure
+        public virtual void AddProperty(String name, KVValue value)
+        {
+            if (IsArray)
+            {
+                //Make up a key for the dictionary
+                Properties.Add(Count.ToString(), value);
+            }
+            else
+            {
+                Properties.Add(name, value);
+            }
+
+            Count++;
+        }
+
+        public void Serialize(StringBuilder stringBuilder, int indent)
+        {
+            if (IsArray)
+            {
+                SerializeArray(stringBuilder, indent);
+            }
+            else
+            {
+                SerializeObject(stringBuilder, indent);
+            }
+        }
+
+        //Serialize the contents of the KV object
+        private void SerializeObject(StringBuilder stringBuilder, int indent)
+        {
+            //Don't enter the top-most object
+            if (indent > 1)
+            {
+                stringBuilder.Append("\n");
+            }
+
+            for (var i = 0; i < indent - 1; i++)
```

Comment:
This indentation-writing loop is everywhere. Consider refactoring it out into a function or extension method.